x-pack/plugins/enterprise_search/public/applications/workplace_search/views/content_sources/components/add_source/add_custom_source/add_custom_source_api.logic.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+
+import { ApiStatus, HttpError } from '../../../../../../../../common/types/api';
+import { flashAPIErrors, clearFlashMessages } from '../../../../../../shared/flash_messages';
+import { HttpLogic } from '../../../../../../shared/http';
+import { AppLogic } from '../../../../../app_logic';
+import { CustomSource } from '../../../../../types';
+
+export interface AddCustomSourceActions {
+  addSource(name: string, baseServiceType?: string): { name: string; baseServiceType: string };
+  addSourceError(code: number, error: string): HttpError;
+  addSourceSuccess(source: CustomSource): CustomSource;
+}
+
+interface AddCustomSourceValues {
+  sourceApi: ApiStatus<CustomSource>;
+}
+
+export const AddCustomSourceApiLogic = kea<
+  MakeLogicType<AddCustomSourceValues, AddCustomSourceActions>
+>({
+  path: ['enterprise_search', 'workplace_search', 'add_custom_source_api_logic'],
+  actions: {
+    addSource: (name, baseServiceType) => ({ name, baseServiceType }),
+    addSourceError: (code, error) => ({ code, error }),
```

Comment:
I like the `addSource` `onAddSourceSuccess` `onAddSourceError` style for actions with a clear hierarchy, and we've been using it in the codebase a lot.  I suspect you wouldn't have written it this way if it wasn't the more correct Redux-y way though.  What's your feeling on this?

Semi-related I wonder if we could take advantage of https://v2.keajs.org/docs/guide/concepts#loaders at some point in the future